### PR TITLE
Cleaned up JSON code and added JSON5 string escape functions

### PR DESCRIPTION
**Pull request by olehermanse**
I want to add more JSON5 support, this is a first step.

### CODE REVIEW

File: tests/unit/json_test.c [hunk 1]
```diff
@@ -1443,6 +1443,32 @@ static void test_parse_object_double_and_trailing_comma(void)
     }
 }
 
+#define assert_json_strings(unescaped, escaped)        \
```
Comment (vpodzime) [inline tests/unit/json_test.c]: Would be nice to add some verb to the name of the macro. Or at least something like `_eq` to the end.

File: libutils/json.h [hunk 2]
```diff
@@ -123,6 +123,21 @@ typedef struct
 char *JsonDecodeString(const char *escaped_string);
 char *JsonEncodeString(const char *const unescaped_string);
 
+typedef struct _Slice
+{
+    // Slice is used to represent a section of memory which may or may not
+    // contain NUL bytes. This is useful for storing the unsecaped versions of
```
Comment (vpodzime) [inline libutils/json.h]: *unescaped*

File: tests/unit/json_test.c [hunk 3]
```diff
@@ -1469,6 +1469,55 @@ static void test_string_escape(void)
         "Hello, world!\\n'Blah', \\\"blah\\\".");
 }
 
+#define assert_json5_data(_size, unescaped, escaped)              \
```
Comment (vpodzime) [inline tests/unit/json_test.c]: Same here with a missing verb in the assert macro names.

File: libutils/json.c [hunk 4]
```diff
@@ -1,7 +1,8 @@
 /*
    Copyright 2019 Northern.tech AS
 
-   This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+   This file is part of CFEngine 3 - written and maintained by Northern.tech
+  AS.
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libutils/json.c]: This is just a bit confusing change in the commit **after** the one changing the line length limit.

[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/json.c]: Ah, seems I fucked up while rebasing.

[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/json.c]: Fixed.

File: libutils/json.c [hunk 5]
```diff
@@ -46,9 +47,8 @@ static const char *const JSON_NULL = "null";
 struct JsonElement_
 {
     JsonElementType type;
-    char *propertyName;         // to avoid having a struct JsonProperty
-    union
-    {
+    char *propertyName; // to avoid having a struct JsonProperty
+    union {
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/json.c]: Adding a commit to the end to fix this, seems like a bug in clang-format.

File: libutils/json.h [hunk 6]
```diff
@@ -154,38 +157,44 @@ void JsonDestroy(JsonElement *element);
 /**
   @brief Destroy a JSON element if needed
   @param element [in] The JSON element to destroy.
-  @param allocated [in] Whether the element was allocated and needs to be destroyed.
+  @param allocated [in] Whether the element was allocated and needs to be
+  destroyed.
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/json.h]: Alright. clang-format doesn't do this automatically, but I can do it manually (and it won't undo it). Fixed in this specific case.

File: libutils/json.h [hunk 7]
```diff
@@ -376,6 +396,43 @@ JsonElement *JsonArrayGet(const JsonElement *array, size_t index);
...
+  @return True if key was removed
+  */
+bool JsonObjectRemoveKey(JsonElement *object, const char *key);
+
+/**
+  @brief Detach json element ownership from parent object;
+  @param object containing the key property
+  @param property name to be detached
+  */
+JsonElement *JsonObjectDetachKey(JsonElement *object, const char *key);
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/json.h]: Yes! Good catch, fixed.

File: tests/unit/json_test.c (hunk 1 above)
Comment (olehermanse) [reply tests/unit/json_test.c]: Alright, added.

File: libutils/json.h (hunk 2 above)
Comment (olehermanse) [reply libutils/json.h]: Thanks, fixed.

File: tests/unit/json_test.c (hunk 3 above)
Comment (olehermanse) [reply tests/unit/json_test.c]: Alright, added.

**Comment by olehermanse** [comment]
@vpodzime Addressed comments, and rebased a few times to move changes to the correct commits. Pleas re-review.

**Pull request merged**